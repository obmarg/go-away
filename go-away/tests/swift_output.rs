#![allow(dead_code)]

use insta::assert_snapshot;

use go_away::{TypeAlias, TypeMetadata, TypeRegistry};

#[derive(TypeMetadata)]
struct MyData {
    field_one: String,
    nested: Nested,
}

#[test]
fn test_struct_output() {
    let mut registry = TypeRegistry::new();
    MyData::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type", content = "data")]
enum AdjacentlyTaggedTupleEnums {
    OptionOne(One),
    OptionTwo(Two),
}

#[derive(TypeMetadata)]
struct Nested {
    #[serde(rename = "some_other_name")]
    a_string: String,
    an_int: i64,
    fulfilment_type: FulfilmentType,
}

#[derive(TypeMetadata)]
enum FulfilmentType {
    Delivery,
    Collection,
}

#[derive(TypeMetadata)]
struct One {
    x: f32,
}

#[derive(TypeMetadata)]
struct Two {
    y: bool,
}

#[test]
fn test_adjacently_tagged_tuple_enums() {
    let mut registry = TypeRegistry::new();
    AdjacentlyTaggedTupleEnums::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}

/*
// TODO: Add StructEnum support...

#[derive(TypeMetadata)]
#[serde(tag = "type", content = "data")]
enum StructEnum {
    OptionOne { x: String, y: i32 },
    OptionTwo { foo: String, bar: Nested },
}

#[test]
fn test_struct_enum() {
    let mut registry = TypeRegistry::new();
    StructEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}
*/

/*

// TODO: Finish this.

#[derive(TypeMetadata)]
#[serde(tag = "type")]
enum InternallyTaggedTupleEnum {
    One(One),
    Two(Two),
}

#[test]
fn test_internally_tagged_tuple_enum() {
    let mut registry = TypeRegistry::new();
    InternallyTaggedTupleEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}
*/

#[derive(TypeMetadata)]
enum ExternallyTaggedTupleEnum {
    One(One),
    Two(Two),
}

#[test]
fn test_externally_tagged_tuple_enum() {
    let mut registry = TypeRegistry::new();
    ExternallyTaggedTupleEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}

#[derive(TypeMetadata)]
struct TypeWithLifetimes<'a, 'b> {
    data: &'a str,
    other: &'b str,
}

#[test]
fn lifetimes_and_strs() {
    let mut registry = TypeRegistry::new();
    TypeWithLifetimes::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}

#[test]
fn type_deduplication() {
    let mut registry = TypeRegistry::new();

    // These both contain `Nested` so there should be one `Nested` type in the output
    //StructEnum::metadata(&mut registry);
    Nested::metadata(&mut registry);
    MyData::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}

#[test]
fn type_aliases() {
    type MyType = std::collections::HashMap<String, i64>;

    let mut registry = TypeRegistry::new();

    MyType::register_alias("MyType", &mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry), @r###"
    typealias MyType = [String: Int64]
    "###);
}
