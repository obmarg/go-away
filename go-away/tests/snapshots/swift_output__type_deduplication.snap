---
source: go-away/tests/swift_output.rs
expression: "go_away::registry_to_output::<go_away::SwiftType>(&registry)"
---
public struct MyData: Hashable, Codable {
    public var fieldOne: String
    public var nested: Nested

    public init(
        fieldOne: String,
        nested: Nested,
    ) {
        self.fieldOne = fieldOne
        self.nested = nested
    }
}

extension MyData {
    enum CodingKeys: String, CodingKey {
        case fieldOne = "field_one"
        case nested = "nested"
    }
}

public struct Nested: Hashable, Codable {
    public var aString: String
    public var anInt: Int64
    public var fulfilmentType: FulfilmentType

    public init(
        aString: String,
        anInt: Int64,
        fulfilmentType: FulfilmentType,
    ) {
        self.aString = aString
        self.anInt = anInt
        self.fulfilmentType = fulfilmentType
    }
}

extension Nested {
    enum CodingKeys: String, CodingKey {
        case aString = "some_other_name"
        case anInt = "an_int"
        case fulfilmentType = "fulfilment_type"
    }
}

public enum FulfilmentType : Codable {
    case delivery
    case collection
    enum CodingKeys: String, CodingKey {
        case delivery = "Delivery"
        case collection = "Collection"
    }
}



